test/e2e/middleware-can-set-the-matcher-in-its-config/index.test.ts

```diff
@@ -0,0 +1,197 @@
+import { createNext } from 'e2e-utils'
+import { NextInstance } from 'test/lib/next-modes/base'
+import { fetchViaHTTP } from 'next-test-utils'
+
+describe('Middleware can set the matcher in its config', () => {
+  let next: NextInstance
+
+  beforeAll(async () => {
+    next = await createNext({
+      files: {
+        'pages/index.js': `
+          export default function Page({ message }) { 
+            return <div>
+              <p>root page</p>
+              <p>{message}</p>
+            </div>
+          } 
+
+          export const getServerSideProps = () => {
+            return {
+              props: {
+                message: "Hello, world."
+              }
+            }
+          }
+        `,
+        'pages/with-middleware/index.js': `
+          export default function Page({ message }) { 
+            return <div>
+              <p>This should run the middleware</p>
+              <p>{message}</p>
+            </div>
+          } 
+
+          export const getServerSideProps = () => {
+            return {
+              props: {
+                message: "Hello, cruel world."
+              }
+            }
+          }
+        `,
+        'pages/another-middleware/index.js': `
+          export default function Page({ message }) { 
+            return <div>
+              <p>This should also run the middleware</p>
+              <p>{message}</p>
+            </div>
+          } 
+
+          export const getServerSideProps = () => {
+            return {
+              props: {
+                message: "Hello, magnificent world."
+              }
+            }
+          }
+        `,
+        'middleware.js': `
+          import { NextResponse } from 'next/server'
+          export const config = {
+            matcher: ['/with-middleware/:path*', '/another-middleware/:path*']
```

Comment:
How about we make this a subset of the `rewrites` syntax in `next.config.js`? That also makes it clear how to add `has` in the future.